File: mpga-plugin/cli/src/commands/evidence.ts
import fs from 'fs';
import path from 'path';
import { Command } from 'commander';
import chalk from 'chalk';
import { log, progressBar } from '../core/logger.js';
import { loadConfig, findProjectRoot } from '../core/config.js';
import { formatEvidenceLink } from '../evidence/parser.js';
import { runDriftCheck, healScopeFile } from '../evidence/drift.js';

export function registerEvidence(program: Command): void {
  const cmd = program
    .command('evidence')
    .description('Evidence link management');

  // evidence verify
  cmd
    .command('verify')
    .description('Check all evidence links resolve to real code')
    .option('--scope <name>', 'Check specific scope only')
    .option('--json', 'Output as JSON')
    .action(async (opts) => {
      const projectRoot = findProjectRoot() ?? process.cwd();
      const config = loadConfig(projectRoot);

      log.info('Verifying evidence links...');
      const report = await runDriftCheck(projectRoot, config.drift.ciThreshold, opts.scope);

      if (opts.json) {
        console.log(JSON.stringify(report, null, 2));
        return;
      }

      log.header('Evidence Verification');

      for (const scope of report.scopes) {
        const icon = scope.healthPct >= 80 ? chalk.green('✓') : scope.healthPct >= 50 ? chalk.yellow('⚠') : chalk.red('✗');
        console.log(`\n${icon} ${chalk.bold(scope.scope)}  ${scope.healthPct}% (${scope.validLinks}/${scope.totalLinks} valid)`);

        if (scope.staleItems.length > 0) {
          log.dim('  Stale links:');
          for (const item of scope.staleItems) {
            log.dim(`    ${formatEvidenceLink(item.link)} — ${item.reason}`);
          }
        }
        if (scope.healedItems.length > 0) {
          log.dim('  Healed links:');
          for (const item of scope.healedItems) {
            log.dim(`    ${item.link.filepath}:${item.newStart}-${item.newEnd} (was ${item.link.startLine}-${item.link.endLine})`);
          }
        }
      }

      console.log('');
      log.bold('Overall');
      console.log(`  Health:  ${progressBar(report.validLinks, report.totalLinks)} (${report.validLinks}/${report.totalLinks})`);
      if (report.totalLinks === 0) {
        log.info('No evidence links found. Run `mpga sync` to generate them.');
      } else if (report.overallHealthPct >= 80) {
        log.success(`Evidence health: ${report.overallHealthPct}%`);
      } else {
        log.warn(`Evidence health: ${report.overallHealthPct}% — run \`mpga evidence heal\` to fix stale links`);
      }
    });

  // evidence heal
  cmd
    .command('heal')
    .description('Re-resolve broken links via AST and update scope files')
    .option('--auto', 'Auto-fix without confirmation')
    .option('--scope <name>', 'Heal specific scope only')
    .action(async (opts) => {
      const projectRoot = findProjectRoot() ?? process.cwd();
      const config = loadConfig(projectRoot);

      log.info('Running evidence heal...');
      const report = await runDriftCheck(projectRoot, config.drift.ciThreshold, opts.scope);

      let totalHealed = 0;
      for (const scope of report.scopes) {
        if (scope.healedItems.length === 0) continue;
        const { healed, content } = healScopeFile(scope);
        if (healed > 0) {
          fs.writeFileSync(scope.scopePath, content);
          log.success(`${scope.scope}: healed ${healed} link(s)`);
          totalHealed += healed;
        }
      }

      const staleCount = report.scopes.reduce((s, r) => s + r.staleLinks, 0);
      if (totalHealed > 0) log.success(`Total healed: ${totalHealed} link(s)`);
      if (staleCount > 0) log.warn(`${staleCount} link(s) could not be healed (symbol not found) — manual review required`);
      if (totalHealed === 0 && staleCount === 0) log.success('All evidence links are already valid.');
    });

  // evidence coverage
  cmd
    .command('coverage')
    .description('Report evidence-to-code ratio')
    .option('--min <pct>', 'Fail if coverage below this %', '20')
    .action(async (opts) => {
      const projectRoot = findProjectRoot() ?? process.cwd();
      const config = loadConfig(projectRoot);
      const report = await runDriftCheck(projectRoot, config.drift.ciThreshold);

      const minPct = parseInt(opts.min);
      log.header('Evidence Coverage');

      for (const scope of report.scopes) {
        const bar = progressBar(scope.validLinks, scope.totalLinks);
        console.log(`  ${scope.scope.padEnd(20)} ${bar}  (${scope.validLinks}/${scope.totalLinks})`);
      }

      console.log('');
      const pct = report.overallHealthPct;
      console.log(`  Overall: ${pct}% (threshold: ${minPct}%)`);

      if (pct < minPct) {
        log.warn(`Coverage ${pct}% is below threshold ${minPct}%`);
        process.exit(1);
      } else {
        log.success(`Coverage ${pct}% meets threshold ${minPct}%`);
      }
    });

  // evidence add
  cmd
    .command('add <scope> <link>')
    .description('Manually add an evidence link to a scope document')
    .action((scopeName: string, link: string) => {
      const projectRoot = findProjectRoot() ?? process.cwd();
      const scopePath = path.join(projectRoot, 'MPGA', 'scopes', `${scopeName}.md`);

      if (!fs.existsSync(scopePath)) {
        log.error(`Scope '${scopeName}' not found.`);
        process.exit(1);
      }

      const content = fs.readFileSync(scopePath, 'utf-8');
      const evidenceLink = link.startsWith('[') ? link : `[E] ${link}`;

      // Insert before the Known unknowns section
      const updated = content.includes('## Known unknowns')
        ? content.replace('## Known unknowns', `${evidenceLink}\n\n## Known unknowns`)
        : content + '\n' + evidenceLink + '\n';

      fs.writeFileSync(scopePath, updated);
      log.success(`Added evidence link to scope '${scopeName}': ${evidenceLink}`);
    });
}

File: mpga-plugin/cli/src/mpga/commands/evidence.py
"""Evidence link management commands."""
from __future__ import annotations

import json
import sys
from pathlib import Path

import click

from mpga.core.config import find_project_root, load_config
from mpga.core.logger import console, log, progress_bar
from mpga.evidence.drift import heal_scope_file, run_drift_check
from mpga.evidence.parser import format_evidence_link

# Evidence health percentage at or above which status is considered good.
EVIDENCE_HEALTH_GOOD_PCT = 80
# Evidence health percentage at or above which a warning (not error) is shown.
EVIDENCE_HEALTH_WARN_PCT = 50


@click.group("evidence")
def evidence() -> None:
    """Evidence link management."""


@evidence.command("verify")
@click.option("--scope", default=None, help="Check specific scope only")
@click.option("--json", "as_json", is_flag=True, help="Output as JSON")
def evidence_verify(scope: str | None, as_json: bool) -> None:
    """Check all evidence links resolve to real code."""
    project_root = find_project_root() or Path.cwd()
    config = load_config(project_root)

    report = run_drift_check(str(project_root), config.drift.ci_threshold, scope)

    if as_json:
        from dataclasses import asdict

        click.echo(json.dumps(asdict(report), indent=2))
        return

    log.header("Evidence Verification")

    for scope_report in report.scopes:
        if scope_report.health_pct >= EVIDENCE_HEALTH_GOOD_PCT:
            icon = "[green]\u2713[/]"
        elif scope_report.health_pct >= EVIDENCE_HEALTH_WARN_PCT:
            icon = "[yellow]\u26a0[/]"
        else:
            icon = "[red]\u2717[/]"
        console.print(
            f"\n{icon} [bold]{scope_report.scope}[/]  "
            f"{scope_report.health_pct}% ({scope_report.valid_links}/{scope_report.total_links} valid)"
        )

        if scope_report.stale_items:
            log.dim("  Stale links:")
            for item in scope_report.stale_items:
                log.dim(f"    {format_evidence_link(item.link)} \u2014 {item.reason}")
        if scope_report.healed_items:
            log.dim("  Healed links:")
            for item in scope_report.healed_items:
                log.dim(
                    f"    {item.link.filepath}:{item.new_start}-{item.new_end} "
                    f"(was {item.link.start_line}-{item.link.end_line})"
                )

    console.print("")
    log.bold("Overall")
    console.print(
        f"  Health:  {progress_bar(report.valid_links, report.total_links)} "
        f"({report.valid_links}/{report.total_links})"
    )
    if report.total_links == 0:
        log.info("No evidence links found. Run `mpga sync` to generate them.")
    elif report.overall_health_pct >= EVIDENCE_HEALTH_GOOD_PCT:
        log.success(f"Evidence health: {report.overall_health_pct}%")
    else:
        log.warn(
            f"Evidence health: {report.overall_health_pct}% \u2014 "
            "run `mpga evidence heal` to fix stale links"
        )


@evidence.command("heal")
@click.option("--auto", "auto", is_flag=True, help="Auto-fix without confirmation")
@click.option("--scope", default=None, help="Heal specific scope only")
def evidence_heal(auto: bool, scope: str | None) -> None:
    """Re-resolve broken links via AST and update scope files."""
    project_root = find_project_root() or Path.cwd()
    config = load_config(project_root)

    log.info("Running evidence heal...")
    report = run_drift_check(str(project_root), config.drift.ci_threshold, scope)

    total_healed = 0
    for scope_report in report.scopes:
        if not scope_report.healed_items:
            continue
        result = heal_scope_file(scope_report)
        if result.healed > 0:
            Path(scope_report.scope_path).write_text(result.content, encoding="utf-8")
            log.success(f"{scope_report.scope}: healed {result.healed} link(s)")
            total_healed += result.healed

    stale_count = sum(r.stale_links for r in report.scopes)
    if total_healed > 0:
        log.success(f"Total healed: {total_healed} link(s)")
    if stale_count > 0:
        log.warn(
            f"{stale_count} link(s) could not be healed (symbol not found) \u2014 "
            "manual review required"
        )
    if total_healed == 0 and stale_count == 0:
        log.success("All evidence links are already valid.")


@evidence.command("coverage")
@click.option("--min", "min_pct", default="20", help="Fail if coverage below this %")
def evidence_coverage(min_pct: str) -> None:
    """Report evidence-to-code ratio."""
    project_root = find_project_root() or Path.cwd()
    config = load_config(project_root)
    report = run_drift_check(str(project_root), config.drift.ci_threshold)

    threshold = int(min_pct)
    log.header("Evidence Coverage")

    for scope_report in report.scopes:
        bar = progress_bar(scope_report.valid_links, scope_report.total_links)
        console.print(
            f"  {scope_report.scope:<20} {bar}  "
            f"({scope_report.valid_links}/{scope_report.total_links})"
        )

    console.print("")
    pct = report.overall_health_pct
    console.print(f"  Overall: {pct}% (threshold: {threshold}%)")

    if pct < threshold:
        log.warn(f"Coverage {pct}% is below threshold {threshold}%")
        sys.exit(1)
    else:
        log.success(f"Coverage {pct}% meets threshold {threshold}%")


@evidence.command("add")
@click.argument("scope_name")
@click.argument("link")
def evidence_add(scope_name: str, link: str) -> None:
    """Manually add an evidence link to a scope document."""
    project_root = find_project_root() or Path.cwd()
    scope_path = project_root / "MPGA" / "scopes" / f"{scope_name}.md"

    if not scope_path.exists():
        log.error(f"Scope '{scope_name}' not found.")
        sys.exit(1)

    content = scope_path.read_text(encoding="utf-8")
    evidence_link = link if link.startswith("[") else f"[E] {link}"

    # Insert before the Known unknowns section
    if "## Known unknowns" in content:
        updated = content.replace(
            "## Known unknowns", f"{evidence_link}\n\n## Known unknowns"
        )
    else:
        updated = content + "\n" + evidence_link + "\n"

    scope_path.write_text(updated, encoding="utf-8")
    log.success(f"Added evidence link to scope '{scope_name}': {evidence_link}")
